keps/sig-apps/20190318-Graduate-CronJob-to-Stable.md

```diff
@@ -0,0 +1,552 @@
+	// hasn't finished yet.
+	ForbidConcurrent ConcurrencyPolicy = "Forbid"
+
+	// ReplaceConcurrent cancels currently running job and replaces it with a new one.
+	ReplaceConcurrent ConcurrencyPolicy = "Replace"
+)
+
+// CronJobConditionType for CronJob conditions
+type CronJobConditionType string
+
+// These are valid conditions of a job.
+const (
+	// CronJobSettled means the cron job controller has
+	CronJobSettled JobConditionType = "Settled"
+)
+
+// CronJobCondition describes a condition state
+type CronJobCondition struct {
+	// Type of job condition, Settled
+	Type CronJobConditionType
+	// Status of the condition, one of True, False, Unknown.
+	Status api.ConditionStatus
+	// Last time the condition was checked.
+	// +optional
+	LastProbeTime metav1.Time
+	// Last time the condition transit from one status to another.
+	// +optional
+	LastTransitionTime metav1.Time
+	// (brief) reason for the condition's last transition.
+	// +optional
+	Reason string
+	// Human readable message indicating details about last transition.
+	// +optional
+	Message string
+}
+
+// CronJobStatus represents the current state of a cron job.
+type CronJobStatus struct {
+	// A list of pointers to currently running jobs.
+	// +optional
+	Active []api.ObjectReference
+
+	// Information when was the last time the job was successfully scheduled.
+	// +optional
+	LastScheduleTime *metav1.Time
+
+	// Information for when was the next time the job will be scheduled.
+    // May not be accurate in some cases for Forbid concurrency policy
+	// +optional
+	NextScheduleTime *metav1.Time
+
+	// Information when was the last time the job successfully completed.
+	// +optional
+	LastSuccessfulTime *metav1.Time
+
+	// Counter for tracking successful Job runs
+	// +optional
+	SuccessfulRuns int64
+
+	// Counter for tracking failed Job runs
+	// +optional
+    FailedlRuns int64
+
+	// Counter for tracking failed Job runs since last successful run
+	// +optional
+    FailuresSinceSuccess int64
+
+	// The set of conditions for this object
+	// +optional
+	Conditions []CronJobCondition
+
+}
+```
+
+## v1beta1 changes
+None
+
+## Validations
+Nothing additional from v1beta1
+
+## Tests
+
+### E2E test
+CronJob E2E test code is [located here](`/test/e2e/apps/cronjob.go`)
+
+#### Existing test cases
+
+- ConcurrencyPolicy
+  - should schedule multiple jobs concurrently when AllowConcurrent
+  - should not schedule new jobs when ForbidConcurrent
+  - should replace jobs when ReplaceConcurrent
+- Suspend
+  - should not schedule jobs when suspended
+- SuccessfulJobsHistoryLimit
+  - should delete successful finished jobs when above successfulJobsHistoryLimit
+- FailedJobsHistoryLimit
+  - should delete failed finished jobs when above failedJobsHistoryLimit
+- Events Recorder
+  - should not emit unexpected warnings
+  - should remove from active list jobs that have been deleted
+
+#### New test cases
+- Schedule *
```

Comment:
This is covered in unit tests, no need for an e2e. Focus with e2e on more of a global state testing.

Reply:
The current controller has a pretty extensive unit suites, I'd expect at least a similar coverage from the new controller.

Reply:
the open bugs (especially https://github.com/kubernetes/kubernetes/issues/86647, https://github.com/kubernetes/kubernetes/issues/83728, https://github.com/kubernetes/kubernetes/issues/82753, https://github.com/kubernetes/kubernetes/issues/74741, https://github.com/kubernetes/kubernetes/issues/63371) against the current controller indicate our unit test coverage is lacking, especially around missed jobs, etc. to me, that indicates a unit testing gap that should be closed as part of this effort

Reply:
I'm not opposed to it, units are limited in what they can effectively test. Missed jobs, and several other use-cases described in the linked issues. Although several of those are already tested:
- https://github.com/kubernetes/kubernetes/blob/fbfbd10c283ea334cdea94f4da946204dfcf88f9/test/e2e/apps/cronjob.go#L233-L253 verifies job cleaning raised in https://github.com/kubernetes/kubernetes/issues/74741
- https://github.com/kubernetes/kubernetes/blob/fbfbd10c283ea334cdea94f4da946204dfcf88f9/test/e2e/apps/cronjob.go#L84-L105 verifies that suspended cronjob is not triggered raised in https://github.com/kubernetes/kubernetes/issues/83728
- https://github.com/kubernetes/kubernetes/issues/63371 is actually solved by adding `StartingDeadlineSeconds`, maybe we should reconsider defaulting that value in `batch/v1`
- https://github.com/kubernetes/kubernetes/blob/fbfbd10c283ea334cdea94f4da946204dfcf88f9/test/e2e/apps/cronjob.go#L194-L231 is covering problem from https://github.com/kubernetes/kubernetes/issues/82753 about deleted job being present in active list

Reply:
Having said that, I want to see a deeper analysis why certain problems describe in those issues are popping up and how to solve them. Example, default `StartingDeadlineSeconds` to solve https://github.com/kubernetes/kubernetes/issues/63371 seems like a viable option. I'd prefer having this explicitly defaulted rather than being done implicitly in the controller.